### PR TITLE
Add ability to filter hooks via simple keys.

### PR DESCRIPTION
**Pull request by tmertens**
Also, add feature spec documentation for various filter types.

This resolves a pain point for me, where we want to create conditional before/after/around blocks that run if a key is set on an example with any value.  Previously, in order to handle any possible value, we had to provide a Proc which evaluates to true for this to work.  This PR adds the ability to simply specify the key as a filter, and if it exists in the metadata hash, it will run the block (provided no other filters filter it out).

For example, suppose we want to timecop an entire example to a specified date/time.  Previously, this would require a Proc to handle evaluation of the metadata value in order to make the block run:

```ruby
config.around(:example, travel: ->(date_time) { date_time.present? }) do |example|
  Timecop.travel example.metadata[:travel] do
    example.run
  end
end
```

This PR makes this scenario simpler and more efficient due to not having to evaluate a Proc for every example containing the specified key:

```ruby
config.around(:example, :travel) do |example|
  Timecop.travel example.metadata[:travel] do
    example.run
  end
end
```

I also added feature specs for all of the filter types, except `:location` and `:ids`, which seem to be more specific to the CLI.  These various options were previously undocumented and only apparent if you actually read through the `Core::MetadataFilter` class or tests.

Necessary Changes:
- Add a filter for when only a key is provided.
- Minor refactor of `Core::MetadataFilter#filter_applies?` to satisfy rubocop, and satisfy this feature.
- Add tests!

All previously existing tests pass without modification, so this should be a backwards compatible feature.

### CODE REVIEW

File: lib/rspec/core/metadata_filter.rb [hunk 1]
```diff
@@ -61,6 +58,14 @@ def line_number_filter_applies?(line_numbers, metadata)
           !(relevant_line_numbers(metadata) & preceding_declaration_lines).empty?
         end
 
+        def proc_filter_applies?(key, proc, metadata)
+          case proc.arity
+          when 0 then proc.call
+          when 2 then proc.call(metadata[key], metadata)
+          else proc.call(metadata[key])
+          end
+        end
```
Comment (myronmarston) [inline lib/rspec/core/metadata_filter.rb]: This extraction is a good refactoring, but is unrelated to the rest of the changes in this commit.  Do you mind putting it in its own commit?

Comment (tmertens) [inline lib/rspec/core/metadata_filter.rb]: sure, I had to pull this out to get rubocop to pass.

File: features/hooks/filtering.feature [hunk 2]
```diff
@@ -244,3 +244,186 @@ Feature: filters
       .after context
       """
 
+  Scenario: Filtering hooks using symbols
```
Comment (myronmarston) [inline features/hooks/filtering.feature]: Playing around with your PR locally, I'm seeing that this scenario is the only new one that requires your changes to pass.  Did you add the others just to flesh this cuke out with more examples or were you thinking they were affected by your changes?

File: lib/rspec/core/metadata_filter.rb [hunk 3]
```diff
@@ -15,22 +15,19 @@ def apply?(predicate, filters, metadata)
         # @private
         def filter_applies?(key, value, metadata)
           silence_metadata_example_group_deprecations do
-            return filter_applies_to_any_value?(key, value, metadata) if Array === metadata[key] && !(Proc === value)
             return location_filter_applies?(value, metadata)          if key == :locations
             return id_filter_applies?(value, metadata)                if key == :ids
             return filters_apply?(key, value, metadata)               if Hash === value
 
             return false unless metadata.key?(key)
+            return true if value.is_a?(TrueClass) && !!metadata[key]
```
Comment (myronmarston) [inline lib/rspec/core/metadata_filter.rb]: Since there are so many values in ruby that are truthy (everything but `false` and `nil`!) I'm concerned about this causing a regression for some users.  An example case I can think of is this: the user is using a `Proc` (which is considered truthy) for their filtering.  Previously, it would have hit the `when Proc` branch below and evaulated the proc to determine if the filter applies.  Here you short circuit that...when in fact, the proc may evaluate to `false`.  

There may be other cases that are potential regressions but that's the only one I can think of.  We should figure out a solution for that.

**Comment by myronmarston** [comment]
:+1: Overall this looks great and I appreciate the thoroughness.

File: lib/rspec/core/metadata_filter.rb (hunk 3 above)
Comment (tmertens) [inline lib/rspec/core/metadata_filter.rb]: This will not be the case, as `value` would be the Proc, which would fail the `is_a?` check.

File: features/hooks/filtering.feature (hunk 2 above)
Comment (tmertens) [inline features/hooks/filtering.feature]: This is to flesh out the cuke.  Not having specific examples here made these filters a (previously) hidden gem of rspec, which are very cool and powerful.  I added the documentation I wish I had when we first started using these filters.

Comment (myronmarston) [inline features/hooks/filtering.feature]: Makes sense.  Do you mind putting them in a separate commit?

It helps keeps the history easier to follow when things like this are committed separately -- otherwise it suggests these scenarios depend upon your changes.

**Comment by tmertens** [comment]
@myronmarston I split up the commits per your request.

File: lib/rspec/core/metadata_filter.rb (hunk 3 above)
Comment (myronmarston) [inline lib/rspec/core/metadata_filter.rb]: You're right :).


Speaking of `is_a?`, we avoid that in RSpec, for the simple reason that users pass all manner of objects to us and we can't necessarily trust that it responds to `is_a?` or that its implementation does what we expect.  For example, `BasicObject` does not respond to `is_a?` or a user may have redefined `is_a?` for their domain (e.g. `user.is_a?(:admin)`).

Instead, we always favor `klass === object` over `object.is_a?(klass)`. In this case, that would be `TrueClass === value`.  If the user is redefining `===` on core built in classes than we're not going to support them (and we've never seen that happen).

Comment (tmertens) [inline lib/rspec/core/metadata_filter.rb]: The bigger concern here IMO is the way `filter_applies_to_any_value?` flips the metadata and filter values when looping back to `filter_applies?`.  This doesn't seem right, but I didn't modify it at all because it seemed risky.  Most cases I could think of though, this doesn't seem to matter, as it should still only match if someone is filtering an array using `before(:foo)` or `before(foo: true)` and there _happens_ to be an array on an example group with the same metadata name and a literal `true` as a member of the array, e.g. `it 'does something', foo: [true] do` but even so, I'm 99.9% certain this would have been previously matched by the string comparison at the bottom of `filter_applies?` so I think this is still not actually a change in behavior.

Comment (tmertens) [inline lib/rspec/core/metadata_filter.rb]: > we always favor klass === object over object.is_a?(klass)

I'll fixup the first commit with this change, tests still pass.

**Comment by tmertens** [comment]
@myronmarston Also split up the doc commits.  Let me know if there are any other changes to be made to get this merged and thanks for the quick review!

**Comment by JonRowe** [comment]
:+1: Thanks @tmertens
